### DISCUSSION TITLE
refactor IfcParse API

### DISCUSSION THREAD

**Discussion by dirkolbrich** [IfcOpenShell/IfcOpenShell#3787] [Ideas]
This is moved from a discussion under the commit https://github.com/IfcOpenShell/IfcOpenShell/commit/3bc9f81b6a2dde6dc33b1976b9d9de9b4bbfb8bc. Replicated and to be continued here as this seems to be a more appropriate place.

To broaden this topic a bit, here are some thoughts I got while working on the `ifcparse/` source:

- put the official express schema files like IFC2X3_TC1.exp together with the generated files under a `ifcparse/schemas/` folder - that way they are separated from the source and you get the direct connection from schema definition to generated file
- move the express-parser tool under is own folder under `tools/express-parser/` outside of the `src/` folder - it is a tool to generate ifcOS source code, not the source itself - also the tool currently is located under `src/ifcopenshell-python/ifcopenshell/express` which is not logical or easy to access
- clean-up the public API of IfcParse to make it a concise scoped library - unify namespaces, make only interfaces or actual code to include in other projects public, everything else should be internal
- unify coding and naming styles (just an optical refactor, but aids in reading the source) - currently its a mix of PascalCase, snake_case wildly mixed between files, cases, functions and variables, sometimes internal variables have a trailing underscore sometimes not, order of public vs private members of times is mixed, spilt up huge files etc.
- the most important IMHO: make the ifcParse library more general useful - currently it only parses files and you initialise a `IfcParse::IfcFile` handle to work with, but what about json streams from a server or db queries or whatever?
here I would imagine a public API of ifcParse like this:
```cpp
#include <ifcopenshell/ifcparse>

IfcParse::Parser parser{};
IfcHierarchy ifcFile = parser->from_file(filepath);
// or
IfcHierarchy ifcStream = parser->from_stream(stream);
```
Sorry for the long post. What do you think and do you have a general direction for this library in mind?

**Comment by dirkolbrich** [comment]
move the express parser tool out of `src/`:

> It's no longer just a static tool for precompilation. You can use it to dynamically load a schema at runtime and parse a file according to that schema. It's also the basis of the express function and where rule validation in ifcopenshell/validate.py

OK, understood.
I am not really versed in Python. After skimming the source of the express_parser tool I found that the include paths for the cpp files are hardcoded - would that be a problem if moving the schema files to a subfolder in term of functionality with Python API?
https://github.com/IfcOpenShell/IfcOpenShell/blob/e3f2f2175df16be52bd7b5617b59cbb79d9e9c33/src/ifcopenshell-python/ifcopenshell/express/templates.py#L29-L35

---

> I don't disagree, but the thing in this community is that all these kind of experiments happen in Python.

Is that a prerequisite? I'm not soo into Python as I prefer a statically typed language. Hope that is not a blocker.

---

> I think the choice we need to make is whether we create this abstraction by virtual functions on the model or by a separation of model and parser. ... Intuitively I would think that virtual functions is a better fit.

I think I do not really grasp the implications of this choice yet. Might have to play with some concepts.

Intuitively I would say that separation of model from parsing step would be better in encapsulation. The model would always be the same regardless of parser. The parser itself has virtual functions for different parser implementations BaseParser -> SqlParser -> EdgeDBParser.

I saw the other day some talks by Greg Schleusner on the NXTBLD conference ([I linked to that in the OsArch forum](https://community.osarch.org/discussion/1124/ifc-merge-and-assemblies)) where he shows his concept of a unified (IFC?) data base using an Entity Component System, where small programs in the AECO cycle only fetch the data they need and play their modification back to the DB. The interesting part is that he mentions they are testing such approaches internally at buildingSmart for a modernization of IFC. Do you have any more information about that?

**Reply by aothms** [reply]
> Is that a prerequisite?

No definitely not. It's just an observation of a preference in the community. In terms of large files, I think C++ is still the only option either by means of HDF5 (python wrapper is incomplete) or RocksDB. I'm still as excited about these things as always, but I'm lately always fighting deadlines for my consultancy stuff.

> The parser itself has virtual functions for different parser implementations BaseParser -> SqlParser -> EdgeDBParser.

Well, we can make this thought experiment maybe a little bit more concrete by just writing some high level methods on these things on paper.

> The interesting part is that he mentions they are testing such approaches internally at buildingSmart for a modernization of IFC. Do you have any more information about that?

I think that would be mostly @tomvandig https://github.com/tomvandig/ecs_ifc_structures and https://github.com/tomvandig/bimledger Unfortunately I've lost track of all the slack and teams workspaces. If you drop me a private note somewhere (e.g info@ifcopenshell.org) I'll connect you with Greg and propose you to represent ifcopenshell in the group (if you'd like that at least).

**Reply by dirkolbrich** [reply]
> I'll connect you with Greg and propose you to represent ifcopenshell in the group

Thank you, but no need to. I'm still in the understanding phase and probably wouldn't be much help currently. I just try to find the bigger picture. The talks by Greg were really inspiring in terms of a unified data structure were you could hook up smaller apps with their own subset of data into.

**Comment by dirkolbrich** [comment]
@aothms I have seen that quote of you in the issue discussion [Strategies on dealing with large IFC datasets](https://github.com/IfcOpenShell/IfcOpenShell/issues/2025):

> it requires a complete rewrite of the parser to store attribute values with static types based on the schema instead of runtime type variants.

Could this be a possible approach for the parser? I will investigate, do you have any conceptual explanation to better understand this?

**Reply by aothms** [reply]
Ok, bare with me, because this is 10y+ of technical debt. The current overview is roughly like this:

```
┌──────────────┐                          ┌───────────────────────────┐       ┌────────────────┐
│              │                          │                           │    *  │                ◄──────────────────────────────────────────────┐
│ IfcBaseClass ├──────────────────────────► IfcEntityInstanceData     ├──────►│ Argument       │                                              │
│              │                          │                           │       │                │◄───────────────┐                             │
└───────▲──▲───┘                          ├───────────────────────────┤       └─▲────▲─────────┘                │                             │
        │  │                              │                           │         │    │                          │                             │
        │  └─────────────────┐            │ id, offset_in_file        │         │    ├───┬─┬──┐                 │                             │
        │                    │            │                           │   ┌─────┴────┴───┤ ├──┴───────────┐ ┌───┴───────────┐         ┌───────┴──────────┐
┌───────┴──────┐    ┌────────┴───────┐    │ declaration* type         │   │              │ │              │ │               │         │                  │
│              │    │                │    │                           │   │ ArgumentList │ │ TokenArgument│ │ EntityArgument│         │ IfcWriteArgument │
│ IfcBaseEntity│    │ IfcBaseType    │    │ Argument** attributes_    │   │              │ │              │ │               │         │                  │
│              │    │                │    │                           │   └──────────────┘ └──────────────┘ └───────────────┘         ├──────────────────┤
└───────▲──────┘    └────────────────┘    └───────────────────────────┘                                                               │                  │
        │                                                                                                                             │ boost::variant...│
        │                                                                                                                             │                  │
┌───────┴──────┐                                                                                                                      └──────────────────┘
│ IfcWall      │
├──────────────┤
│              │
│ GlobalId()   │
│              │
│              │
│              │
│              │
│              │
│              │
│              │
└──────────────┘
```

Aim for all of this was quickly scanning through the model, lazily and partially apply a schema to that. Parts of that made sense 10y ago (most of the files were largely invalid, you would only read partial data). But I think (it's hard to measure) that these days there is a considerable overhead to this.

I eliminated one thing already some time ago. Previously we had std::vector<Argument*> (which is 3 pointers for the vector + 1 pointer for each attribute value). I changed that to Argument** to have 1 pointer for the array (we don't need to store the length because we can derive that from the declaration*, just requires some extra bookkeeping during parse and we can no longer retain values outside of the length of values allowed by the type).

But what if we take this one step further? We replace this IfcEntityInstanceData with `std::tuple<std::string, IfcOwnerHistory*, ...>` each specifically for the type. We would basically define something like:

```C++
template <typename T>
struct instance_data {};

template <>
struct instance_data<Ifc2x3::IfcWallStandardCase> {
  typedef std::tuple<std::string, IfcOwnerHistory*, ...> type;
}
```

Not only would that shave off the last pointer for the Argument** array, but also eliminate all the pointers for respective attribute values.

And that's nice in terms of memory usage, but even better for performance due to less pointer chasing and better cache coherency.

This is a huge effort, also because it completely changes the parser, python bindings, validator. But I think today with modern C++ it's much more feasible than when we started.

**Reply by aothms** [reply]
PS There's additional benefits regarding this wrt to how aggregates are stored (very inefficiently retrieved). And it would also mean that we can have IfcRoot::GlobalId return `const std::string&` reference because we can a return a reference to std::get<0>(data_);. Now we need to make copies, also for e.g CartesianPoint::Coordinates.

**Reply by dirkolbrich** [reply]
Thank you for this explanation. I have to confess that I don't understand this fully as I lack the deeper knowledge of C++ template programming and need to fully grasp the parsing part of IFC yet. My mind just left the chat... I probably need to build myself some simple experiments first.

**Comment by dirkolbrich** [comment]
> Well, we can make this thought experiment maybe a little bit more concrete by just writing some high level methods on these things on paper.

OK, I did some deeper investigation and reading of the /ifcparse library. And I'd like to propose an API rewrite/change. 

But first, the ->

## Why?

IfcOpenShell and its base libraries are fantastic and very useful. That's proven by the use within the BlenderBim add-on and the numerous little tools in its repo. **BUT**, the library can be better.

As a new user of the library it is not easy to understand the public API of IfcParse or IfcGeom. Where is the obvious entry point? Which classes do I use to get started? What do all the public available classes of the libraries do? What are the different classes of "aggregate_of" for, what is their purpose? 

To grasp the functionality of the libraries one must dig a bit too deep to get a good start. And speaking frankly, after a deeper engagement with the code it is quite messy.

So, let's start with a ->

## Vision

I would like to _"Make IfcOpenShell THE benchmark library for handling IFC schemas"_.

What makes a good library?

- clear and concise API
- clear code structure
- developer experience - documentation, examples, onboarding
- tested and benchmarked
- versatile and expandable

And, here is my ->

## Proposal

_(I will cut the prose short and just list some ideas)_

- all namespaces lower case -> in accordance with the standard library
- ifcparse, ifcserialize, ifcgeom
- rename IfcUtil to ifcbase namespace - these are the base classes of the IfcSchema
- new ifcutil namespace for utilities like file handlers, databases, streams etc.
- add an ifcschema namespace to group the generated Ifc classes and hold the IfcHierarchy(naming?)
- all classes within a namespace
- Naming style: all classes PascalCase -> consistency throughout the source code
- only leave the API entry points public, everything else goes private/internal
- instead of using the specific class IfcParse::IfcFile use a generic IfcResource handle

Here is what I think would be a starting point on how to use the API:

```cpp
#include <ifcopenshell/ifcgeom>
#include <ifcopenshell/ifcparse>
#include <ifcopenshell/ifcschema> // <-- find a way to not need including these 
#include <ifcopenshell/ifcutil>   // <--

// create a parser and obtain a file handle
ifcparse::Parser parser{};
ifcutil::IfcResource ifc_file{filepath};
// ifcutil::IfcResource ifc_stream{stream}; //  

// the parser uses the resource to obtain the IFC hierarchy tree
ifcschema::IfcHierarchy ifc_hierarchy = ifc_parser->parse(ifc_file);

ifcgeom::Element wall_01 = ifc_hierarchy->get_element<IfcWall>.first();
// do something with the wall

ifc_hierarchy->attach(wall_01).to(ifc_hierarchy->get_element<IfcStorey>.by_name("Ground Floor"));
```

And now some first mock-up of used namespaces, classes and the interaction between these:

```mermaid
graph LR

%% namespace ifcutil
subgraph ifcutil
    resource(ifcutil::IfcResource)
    file(ifcutil::IfcFile)
        filereader(ifcutil:FileReader)
        filewriter(ifcutil:FileWriter)
    db(ifcutil::IfcDB)
    stream(ifcutil::IfcStream)

    filereader & filewriter --> file
    file & db & stream --> resource
end

%% namespace ifcparse
subgraph ifcparser
    parser(ifcparse::Parser)
    spfparser(ifcparse::SpfParser)

    resource --> spfparser --> parser
end

%% namespace ifcserialize
subgraph ifcserializer
    serializer(ifcserialize::Serializer)
    spfserializer(ifcserialize::SpfSerializer)

    serializer --> spfserializer --> resource
end

%% namespace ifcschema
subgraph ifcschema
    schema(ifcschema::IfcSchema)
    hierarchy(ifcschema::IfcHierarchy)
    2x3(ifcschema::Ifc2x3)
    4x0(ifcschema::Ifc4)
    4x3(ifcschema::Ifc4x3)
    5x0(ifcschema::Ifc5)

    subgraph Ifc4x3
        root(ifcschema::Ifc4x3::IfcRoot)
        wall(ifcschema::Ifc4x3::IfcWall)
        other(ifcschema::Ifc4x3::Ifc...)
    end

    2x3 & 4x0 & 4x3 --> schema
    5x0 -.-> schema
    hierarchy -- has --> schema
    hierarchy ---> root
    root --> wall
    root -.-> other
end

%% namespace ifcbase
subgraph ifcbase
    baseinterface(ifcbase::BaseInterface)
    baseclass(ifcbase::BaseClass)
    baseentity(ifcbase::BaseEntity)
    basetype(ifcbase::BaseType)

    baseentity & basetype --> baseclass --> baseinterface 
end

%% namespace geom
subgraph ifcgeom
    iterator(ifcgeom::Iterator)
    element(ifcgeom::Element)
    representation(ifcgeom::Representation)
    filter(ifcgeom::Filter)

    element -.-> iterator
end

%% connections between namespaces
schema -- "used by" ---> parser
parser -- "parses into" ---> hierarchy
hierarchy ---> serializer
baseentity ---> wall
wall ---> element
iterator -- "iterates over hierarchy to get element" --> hierarchy
```

I hope the mermaid.js graph renders OK (at least it does in Obsidian), else have a look at the source code block.

**Reply by aothms** [reply]
![afbeelding](https://github.com/IfcOpenShell/IfcOpenShell/assets/1096535/6050f132-af42-4739-b6db-fd26d1fa6c29)

**Reply by aothms** [reply]
Thanks for keeping the ball rolling 👍 

> all namespaces lower case -> in accordance with the standard library

+1

> ifcparse, ifcserialize, ifcgeom

how do you feel about:

ifcopenshell::model ifcopenshell::geometry::serializers ifcopenshell::geometry

for the first, I find it hard to delineate currently model and parser (we talked about this). Otherwise maybe ifcopenshell::data?

I would also be ok with shortening the ifcopenshell namespace to just ifc (like you said: THE benchmark library for handling IFC schemas). Or we could leave it to the user and recommend something like `namespace ifc = ifcopenshell` in client code.

I prefer for library names no longer to use any ifc prefix to really prevent any confusion between code and schema classes (especially if your propose CamelCase). Because we wrap everything in a root namespace, this should be possible.

> rename IfcUtil to ifcbase namespace - these are the base classes of the IfcSchema

I would merge it with ifcopenshell::model/ifcopenshell::data. They are not really utilities. These 2 to 3 abstract base classes don't need a separate namespace. It's also the question what's their future role once we've understood the relationship between parser and model.

> new ifcutil namespace for utilities like file handlers, databases, streams etc.

I think util(ities) ought to be convenience or supplementary things. Not the core things like handlers, databases, streams. Maybe ifcopenshell::data or just ifcopenshell::core?

> add an ifcschema namespace to group the generated Ifc classes and hold the IfcHierarchy(naming?)

I agree, but how about ifcopenshell::schemas::ifc2x3?

> all classes within a namespace
>
> Naming style: all classes PascalCase -> consistency throughout the source code
>
> only leave the API entry points public, everything else goes private/internal
>
> instead of using the specific class IfcParse::IfcFile use a generic IfcResource handle

I'm quite fond of something like ifcopenshell::internal / ifcopenshell::detail or something like to mark certain classes/functors as outside of the public API.

Instead of IfcResource here I prefer ifcopenshell::model or ifcopenshell::document (CamelCase if need be).

The difficult thing is that nearly every word is overloaded in IFC. Documents exist as IfcDocumentReference, resources exist as the resource layer (IfcGeometryResource subschema, ...) vs the domain schemas. What word is still left to represent the abstract interface for a model? The STEP people use "population" a lot, which I find strange, but at least it's not something that exist in the IFC discourse.

> `ifc_hierarchy->attach(wall_01).to(ifc_hierarchy->get_element<IfcStorey>.by_name("Ground Floor"));`

In nearly all cases it's safe to use a generic verb like `attach()` and given both ends of the relationship you know if you're talking about containment, aggregation, property set association, etc. But there are also cases where you really need to specify the relationship type you'd be creating. So it might be safer to have attach as a:

```C++
template <typename From, typename To, typename Relationship=default_relationship<From, To>::type>
void attach(From from, To to) { ...
```

`link()` might be a more general word than `attach()` which might give people too strong of a physical attachment idea.

This example does reinforce the idea that maybe we should indeed give filter a more central place instead of just geometry iterator. Why not `ifcopenshell::make_filter().by_name("Ground Floor").by_type<IfcBuildingStorey>()(ifc_hierarchy).first().` instead of `ifc_hierarchy->get_element<IfcStorey>.by_name("Ground Floor")`. And use the same approach for geometry iteration:

```C++
for (auto elem : ifcopenshell::make_filter().by_type<IfcWallStandardCase>()(model)) {
    serialize(elem);
}
```

(or use something like std::copy() and boost transform_iterator)

### other remarks

I like how you positioned the IfcHierarchy. But maybe this could become `ifcopenshell::model`? Like a richer interface on top of just a `file` as a set of instances? It also make sense in terms of the iterator (maybe it could just become ifcopenshell::model::begin() / end()). I.e an actual iterator so that you could just use a range based for loop. And maybe filter is a just a function you apply on a model that yields another model? So that you can also meaningfully compose them?

I think this is also a moment to bring in this discussion: https://community.osarch.org/discussion/comment/17541/#Comment_17541 The performance of making changes to relationships is currently really poor. What does this mean for the new API? How can we make changes to edges cheaper while still adhering to the IFC schema?

**Comment by dirkolbrich** [comment]
> I would also be ok with shortening the ifcopenshell namespace to just ifc ... 

Yes, I like shortening the namespace even further too. An `ifc` namespace is IMV too generic, as there are other libraries out there which use IFC, the library namespace should make the used library visible. Maybe instead use `ifcos` (abbreviation for IfcOpenShell) as the public namespace?

```cpp
#include <ifcos/parse>

ifcos::Parser parser{};
```

The user could include different libraries `<ifcos/parse>`, `<ifcos/serialize>`, `<ifcos/geom>` according to use case, but there should be no further/additional namespace for the public API - just like the `std` library -> maybe only for `ifcos::schema::Ifc2x3`? This would keep a short namespace consistent across the different library calls.

> I find it hard to delineate currently model and parser

Yes, the current API with `ifcparse::` does not really provide a distinction between the in-memory IFC hierarchy and parser.

My main idea is to define a concise core functionality of the IfcOS cpp library (with corresponding Python bindings) and to group the subelements of this under `/core` within the source tree. I see the following `core` functionality:

- resource - for files, databases, streams etc. anything where ifc data is saved and read from/written to, in short the persistent source of data
    - each of these storage forms has its own specialised read/write functionality according to medium
    - the grouping `ifcutil` for this is indeed not right as it describes supplementary things
    - I would not use `data` as well, as the storage mediums are not the data in itself, but the resource of data
    - `core` is it neither -> core is in my eyes the complete base functionality of IfcOS to provide a valid IFC hierarchy model someone can work with
- parser - for _reading_ data from source and parsing it into a valid IFC hierarchy model
- serialize - for _writing_ the IFC hierarchy model into data (spf, json, sql etc.) to be saved to a source medium
- schema - the schema definitions 2x3, 4, 4x3, 5 ...
- hierarchy (model? core? tree? i don't know yet) - the actual IFC hierarchy tree in memory with which other apps like `/geom` or `4D` can work with
    - currently hierarchy is the most fitting description for me, as all the different IFC elements, properties and relations are hierarchically ordered under an IfcRoot (I might be wrong in understanding this correctly)

All these 5 base functionalities form the core library together, but the use of `/core` is restricted to only separate the base functionality within the source folder:

```shell
/src
- ...
- ifcgeom
- ifccore
    - resource
    - parse
    - serialize
    - schema
    - hierarchy
- ...
```

Note: After some more thought about this I start to like the idea of a `core` library:

```cpp
#include <ifcos/core>

ifcos::Parser parser{};
```

Even thought name `core` does not really convey the functionality of the library. Semantically it feels a bit too undefined - what is `core`, what not?

> I agree, but how about `ifcopenshell::schemas::ifc2x3`?

I would use the singular form:
```cpp
ifcos::schema::ifc2x3
```

> The difficult thing is that nearly every word is overloaded in IFC. Documents exist as IfcDocumentReference, resources exist as the resource layer (IfcGeometryResource subschema, ...) vs the domain schemas.

In good Domain Driven Design tradition I would choose a wording which is as close to the actual business domain as possible. In this case as close to IFC as it gets. This ensures that even non-programmers are somewhat familiar with the meaning of library functionality and its interconnection between different classes/functions.

> Instead of IfcResource here I prefer ifcopenshell::model or ifcopenshell::document (CamelCase if need be).

I disagree.
_Resource_ is the source of the data, the file or the db. Just like in the IFC schema, but here it is the "Data Resource layer". So using this naming would actually be consistent with the domain.
_Model_ is more the mental model of the IFC hierarchy tree -> so better name it hierarchy directly. 
_Document_ is more like the output from a model, like a plan document or a schedule document. It can be a resource to, but the word document restricts that too much to a more "paperlike" concept. Here i would keep it more generic to anything that can hold data outside of the in-memory hierarchy model.

> What word is still left to represent the abstract interface for a model?

Here we talk about the complete IFC hierarchy parsed from a file (or database) ready to be used, filtered, queried, manipulated - correct? Then it is hierarchy, I think.

Aside: Do you have a high level overview of what data structure is currently constructed in memory by ifcParse from the parsed file? 

> The STEP people use "population" a lot, which I find strange, but at least it's not something that exist in the IFC discourse.

I find that strange too. And I wouldn't use it as it is a domain foreign concept. It is not connected to an actual population in the IFC hierarchy.

> `link()` might be a more general word than `attach()` which might give people too strong of a physical attachment idea.

`link` is a good word if you have a separate element which is somehow connected to other elements, but still can exist individually. `attach` does indeed resembles a strong physical connection, but isn't that actually very aptly expressed, as a wall is sorted within the IFC hierarchy to be located under a storey?
Anyway, these are wording problems to describe interaction between elements. Basically the functionality naming on top of the base hierarchy. I think these can wait until after the basic structure. 

> ```cpp
> ifcopenshell::make_filter().by_name("Ground Floor").by_type<IfcBuildingStorey>()(ifc_hierarchy).first()
> ```

This is cool. A syntax of chainable basic functions. With the API adjustments from above this would read like (static function, right!?):

```cpp
ifcos::IfcWall my_wall = ifcos::filter().by_name("Ground Floor").by_type<ifcos::IfcWall>().first()(ifc_hierarchy);
```

I also like the idea of:
> filter is a just a function you apply on a model that yields another model

This way if you do a transformation to the hierarchy you can get back a valid sub hierarchy, ready to be saved or transfered. -> think of atomic change to the main IFC model in the cloud.

**Comment by dirkolbrich** [comment]
> What are the different classes of "aggregate_of" for, what is their purpose?

Just as a reminder for myself linking this short Q&A here, as I slowly start to understand `aggregate_of`: https://github.com/IfcOpenShell/IfcOpenShell/discussions/3737
(where I currently think, this functions should not be part of the public API)

**Comment by dirkolbrich** [comment]
I started to think of a possible refactored classes structure. Will see if I can get a basic API based on this to work.
![IfcOpenShell_core_graph_v01 png](https://github.com/IfcOpenShell/IfcOpenShell/assets/3765119/b9d066d2-7afc-4d68-9e69-8339ee415d58)

**Reply by aothms** [reply]
My struggle is still with Resource - Model. Does it really make sense to talk about parser and serializer in the context of a database? I think the Resource needs to be more leaky or else we aren't able to exploit the inherent advantages of them. There are people that want to convert 50GB IFC models. They currently can't because of memory constraints. This needs to be solved somewhere between resource and model.

![afbeelding](https://github.com/IfcOpenShell/IfcOpenShell/assets/1096535/3c5666d1-a45a-4808-b4ec-fc2fd5daf762)

**Comment by dirkolbrich** [comment]
Long time, no see. So I had time to think more about a possible structure which makes heavy use of interfaces for a more modular approach of the API. I'll just dump the graph here. Will write down some thoughts on this later, when returned into the mental model of IfcOS.
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/3765119/d66fa084-b437-4ec8-86f9-b64621bf5658)
